c++/src/kj/compat/tls.c++

```diff
@@ -445,8 +457,53 @@ public:
   }
 
 private:
+  Promise<void> onAcceptSuccess(AuthenticatedStream&& stream) {
+    auto acceptPromise = kj::evalNow([&] {
+      // Do the SSL acceptance procedure.
+      return tls.wrapServer(kj::mv(stream));
+    });
+
+    auto sslPromise = acceptPromise.then([this](auto&& stream) -> Promise<void> {
+      // This is only attached to the success path, thus the error handler will catch if our
+      // promise fails.
+      queue.push(kj::mv(stream));
+      return kj::READY_NOW;
+    });
+    tasks.add(kj::mv(sslPromise));
+
+    // Queue up the next accept loop immediately without waiting for SSL_accept()/wrapServer().
+    return acceptLoop();
+  }
+
+  void onAcceptFailure(Exception&& e) {
+    KJ_LOG(ERROR, "TlsConnectionReceiver could not accept()", e);
+
+    // Store this exception to reject all future calls to accept() and reject any unfulfilled
+    // promises from the queue.
+    maybeInnerException = kj::mv(e);
+    queue.rejectAll(Exception(KJ_REQUIRE_NONNULL(maybeInnerException)));
+  }
+
+  Promise<void> acceptLoop() {
+    // Accept one connection and queue up the next accept on our TaskSet.
+
+    return inner->acceptAuthenticated().then(
```

Comment:
If `inner->acceptAuthenticated()` throws synchronously, it looks like the exception won't be caught and will bubble up to the TaskSet, causing `taskFailed()` to be invoked, which merely logs the exception. The loop will be broken but the application isn't notified so it'll hang forever.

I think to solve this, the `acceptLoop()` promise should _not_ go in the `TaskSet`, but should instead be stored as a separate member `Promise<void> acceptLoopTask`, and should have an explicit error handler like:

```
acceptLoopTask = acceptLoop().eagerlyEvaluate([this](kj::Exception&& e) {
  onAcceptFailure(kj::mv(e));
});
```

Then you can also remove the async error handler here, since you can rely on the top-level handler to take care of both sync and async exceptions.

Reply:
I think I like this way better overall. Thanks!

I should ask in this context: What is the criteria for a Promise to be "waiting"? I could see several different criteria, and I've been nervous that keeping a `Promise` member would not count.

Reply:
A `Promise` that is just sitting as a member will execute, with one caveat: trailing `.then()` continuations that only transform the result (they do not return another Promise) will be executed lazily when the result is actually needed. However, using `.eagerlyEvaluate()` prevents this. So the general rule of thumb is: if you're going to keep a `Promise` around, make sure to call `.eagerlyEvaluate()` on it. This also forces you to provide an exception handler, so that errors aren't silently ignored.